### ISSUE TITLE
Link error (undefined abseil reference) in debug mode

### ISSUE THREAD

**Issue by e8y** [protocolbuffers/protobuf#17404]
**What version of protobuf and what language are you using?**
Version: v27.2
Language: C++

**What operating system (Linux, Windows, ...) and version?**

Ubuntu 22.04 LTS (jammy) in a Docker container

**What runtime / compiler are you using (e.g., python version or gcc version)**

LLVM clang 18

**What did you do?**

See the code to reproduce the compile error [in this repo](https://github.com/e8y/protobuf-bug-report/tree/2024_07_10).

```
$ ./launch.sh
```

**What did you expect to see**

Successful compilation, which does happen when you use Release mode, as in:

```
$ ./launch.sh --interactive
# ./reproduce.sh Release
```

**What did you see instead?**

   ```
   [1/1] Linking CXX executable main
   FAILED: main
   : && /usr/bin/clang++-18 -g  CMakeFiles/messages.dir/generated/src/messages.pb.cc.o CMakeFiles/main.dir/src/main.cpp.o -o main  /opt/protobuf/lib/libprotobuf.a  /opt/protobuf/lib/libabsl_log_internal_check_op.a  /opt/protobuf/lib/libabsl_leak_check.a  /opt/protobuf/lib/libabsl_die_if_null.a  /opt/protobuf/lib/libabsl_log_internal_conditions.a  /opt/protobuf/lib/libabsl_log_internal_message.a  /opt/protobuf/lib/libabsl_log_internal_nullguard.a  /opt/protobuf/lib/libabsl_examine_stack.a  /opt/protobuf/lib/libabsl_log_internal_format.a  /opt/protobuf/lib/libabsl_log_internal_proto.a  /opt/protobuf/lib/libabsl_log_internal_log_sink_set.a  /opt/protobuf/lib/libabsl_log_sink.a  /opt/protobuf/lib/libabsl_log_entry.a  /opt/protobuf/lib/libabsl_flags_internal.a  /opt/protobuf/lib/libabsl_flags_marshalling.a  /opt/protobuf/lib/libabsl_flags_reflection.a  /opt/protobuf/lib/libabsl_flags_config.a  /opt/protobuf/lib/libabsl_flags_program_name.a  /opt/protobuf/lib/libabsl_flags_private_handle_accessor.a  /opt/protobuf/lib/libabsl_flags_commandlineflag.a  /opt/protobuf/lib/libabsl_flags_commandlineflag_internal.a  /opt/protobuf/lib/libabsl_log_initialize.a  /opt/protobuf/lib/libabsl_log_internal_globals.a  /opt/protobuf/lib/libabsl_log_globals.a  /opt/protobuf/lib/libabsl_vlog_config_internal.a  /opt/protobuf/lib/libabsl_log_internal_fnmatch.a  /opt/protobuf/lib/libabsl_raw_hash_set.a  /opt/protobuf/lib/libabsl_hash.a  /opt/protobuf/lib/libabsl_city.a  /opt/protobuf/lib/libabsl_low_level_hash.a  /opt/protobuf/lib/libabsl_hashtablez_sampler.a  /opt/protobuf/lib/libabsl_random_distributions.a  /opt/protobuf/lib/libabsl_random_seed_sequences.a  /opt/protobuf/lib/libabsl_random_internal_pool_urbg.a  /opt/protobuf/lib/libabsl_random_internal_randen.a  /opt/protobuf/lib/libabsl_random_internal_randen_hwaes.a  /opt/protobuf/lib/libabsl_random_internal_randen_hwaes_impl.a  /opt/protobuf/lib/libabsl_random_internal_randen_slow.a  /opt/protobuf/lib/libabsl_random_internal_platform.a  /opt/protobuf/lib/libabsl_random_internal_seed_material.a  /opt/protobuf/lib/libabsl_random_seed_gen_exception.a  /opt/protobuf/lib/libabsl_statusor.a  /opt/protobuf/lib/libabsl_status.a  /opt/protobuf/lib/libabsl_cord.a  /opt/protobuf/lib/libabsl_cordz_info.a  /opt/protobuf/lib/libabsl_cord_internal.a  /opt/protobuf/lib/libabsl_cordz_functions.a  /opt/protobuf/lib/libabsl_exponential_biased.a  /opt/protobuf/lib/libabsl_cordz_handle.a  /opt/protobuf/lib/libabsl_crc_cord_state.a  /opt/protobuf/lib/libabsl_crc32c.a  /opt/protobuf/lib/libabsl_crc_internal.a  /opt/protobuf/lib/libabsl_crc_cpu_detect.a  /opt/protobuf/lib/libabsl_bad_optional_access.a  /opt/protobuf/lib/libabsl_strerror.a  /opt/protobuf/lib/libabsl_str_format_internal.a  /opt/protobuf/lib/libabsl_synchronization.a  /opt/protobuf/lib/libabsl_stacktrace.a  /opt/protobuf/lib/libabsl_symbolize.a  /opt/protobuf/lib/libabsl_debugging_internal.a  /opt/protobuf/lib/libabsl_demangle_internal.a  /opt/protobuf/lib/libabsl_graphcycles_internal.a  /opt/protobuf/lib/libabsl_kernel_timeout_internal.a  /opt/protobuf/lib/libabsl_malloc_internal.a  /opt/protobuf/lib/libabsl_time.a  /opt/protobuf/lib/libabsl_civil_time.a  /opt/protobuf/lib/libabsl_time_zone.a  /opt/protobuf/lib/libabsl_bad_variant_access.a  /opt/protobuf/lib/libutf8_validity.a  /opt/protobuf/lib/libabsl_strings.a  /opt/protobuf/lib/libabsl_int128.a  /opt/protobuf/lib/libabsl_strings_internal.a  /opt/protobuf/lib/libabsl_string_view.a  /opt/protobuf/lib/libabsl_base.a  /opt/protobuf/lib/libabsl_spinlock_wait.a  -lrt  /opt/protobuf/lib/libabsl_throw_delegate.a  /opt/protobuf/lib/libabsl_raw_logging_internal.a  /opt/protobuf/lib/libabsl_log_severity.a && :
   /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x23
   CMakeFiles/messages.dir/generated/src/messages.pb.cc.o: in function `absl::lts_20240116::log_internal::LogMessage::operator<<(unsigned long)':
   messages.pb.cc:(.text._ZN4absl12lts_2024011612log_internal10LogMessagelsEm[_ZN4absl12lts_2024011612log_internal10LogMessagelsEm]+0x19): undefined reference to `_ZN4absl12lts_2024011612log_internal10LogMessagelsImTnNSt9enable_ifIXntsr4absl16HasAbslStringifyIT_EE5valueEiE4typeELi0EEERS2_RKS5_'
   clang++-18: error: linker command failed with exit code 1 (use -v to see invocation)
   ninja: build stopped: subcommand failed.
   ```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment**

The issue can be reproduced with a clean container after installing LLVM 18 and Protobuf 27.2. (I also tried this on Protobuf 26.1 to the same result). The test code uses CMake.

**Comment by e8y** [comment]
The link command for Release, which is successful:

```
: && /usr/bin/clang++-18 -O3 -DNDEBUG  CMakeFiles/messages.dir/generated/src/messages.pb.cc.o CMakeFiles/main.dir/src/main.cpp.o -o main  /opt/protobuf/lib/libprotobuf.a  /opt/protobuf/lib/libabsl_log_internal_check_op.a  /opt/protobuf/lib/libabsl_leak_check.a  /opt/protobuf/lib/libabsl_die_if_null.a  /opt/protobuf/lib/libabsl_log_internal_conditions.a  /opt/protobuf/lib/libabsl_log_internal_message.a  /opt/protobuf/lib/libabsl_log_internal_nullguard.a  /opt/protobuf/lib/libabsl_examine_stack.a  /opt/protobuf/lib/libabsl_log_internal_format.a  /opt/protobuf/lib/libabsl_log_internal_proto.a  /opt/protobuf/lib/libabsl_log_internal_log_sink_set.a  /opt/protobuf/lib/libabsl_log_sink.a  /opt/protobuf/lib/libabsl_log_entry.a  /opt/protobuf/lib/libabsl_flags_internal.a  /opt/protobuf/lib/libabsl_flags_marshalling.a  /opt/protobuf/lib/libabsl_flags_reflection.a  /opt/protobuf/lib/libabsl_flags_config.a  /opt/protobuf/lib/libabsl_flags_program_name.a  /opt/protobuf/lib/libabsl_flags_private_handle_accessor.a  /opt/protobuf/lib/libabsl_flags_commandlineflag.a  /opt/protobuf/lib/libabsl_flags_commandlineflag_internal.a  /opt/protobuf/lib/libabsl_log_initialize.a  /opt/protobuf/lib/libabsl_log_internal_globals.a  /opt/protobuf/lib/libabsl_log_globals.a  /opt/protobuf/lib/libabsl_vlog_config_internal.a  /opt/protobuf/lib/libabsl_log_internal_fnmatch.a  /opt/protobuf/lib/libabsl_raw_hash_set.a  /opt/protobuf/lib/libabsl_hash.a  /opt/protobuf/lib/libabsl_city.a  /opt/protobuf/lib/libabsl_low_level_hash.a  /opt/protobuf/lib/libabsl_hashtablez_sampler.a  /opt/protobuf/lib/libabsl_random_distributions.a  /opt/protobuf/lib/libabsl_random_seed_sequences.a  /opt/protobuf/lib/libabsl_random_internal_pool_urbg.a  /opt/protobuf/lib/libabsl_random_internal_randen.a  /opt/protobuf/lib/libabsl_random_internal_randen_hwaes.a  /opt/protobuf/lib/libabsl_random_internal_randen_hwaes_impl.a  /opt/protobuf/lib/libabsl_random_internal_randen_slow.a  /opt/protobuf/lib/libabsl_random_internal_platform.a  /opt/protobuf/lib/libabsl_random_internal_seed_material.a  /opt/protobuf/lib/libabsl_random_seed_gen_exception.a  /opt/protobuf/lib/libabsl_statusor.a  /opt/protobuf/lib/libabsl_status.a  /opt/protobuf/lib/libabsl_cord.a  /opt/protobuf/lib/libabsl_cordz_info.a  /opt/protobuf/lib/libabsl_cord_internal.a  /opt/protobuf/lib/libabsl_cordz_functions.a  /opt/protobuf/lib/libabsl_exponential_biased.a  /opt/protobuf/lib/libabsl_cordz_handle.a  /opt/protobuf/lib/libabsl_crc_cord_state.a  /opt/protobuf/lib/libabsl_crc32c.a  /opt/protobuf/lib/libabsl_crc_internal.a  /opt/protobuf/lib/libabsl_crc_cpu_detect.a  /opt/protobuf/lib/libabsl_bad_optional_access.a  /opt/protobuf/lib/libabsl_strerror.a  /opt/protobuf/lib/libabsl_str_format_internal.a  /opt/protobuf/lib/libabsl_synchronization.a  /opt/protobuf/lib/libabsl_stacktrace.a  /opt/protobuf/lib/libabsl_symbolize.a  /opt/protobuf/lib/libabsl_debugging_internal.a  /opt/protobuf/lib/libabsl_demangle_internal.a  /opt/protobuf/lib/libabsl_graphcycles_internal.a  /opt/protobuf/lib/libabsl_kernel_timeout_internal.a  /opt/protobuf/lib/libabsl_malloc_internal.a  /opt/protobuf/lib/libabsl_time.a  /opt/protobuf/lib/libabsl_civil_time.a  /opt/protobuf/lib/libabsl_time_zone.a  /opt/protobuf/lib/libabsl_bad_variant_access.a  /opt/protobuf/lib/libutf8_validity.a  /opt/protobuf/lib/libabsl_strings.a  /opt/protobuf/lib/libabsl_int128.a  /opt/protobuf/lib/libabsl_strings_internal.a  /opt/protobuf/lib/libabsl_string_view.a  /opt/protobuf/lib/libabsl_base.a  /opt/protobuf/lib/libabsl_spinlock_wait.a  -lrt  /opt/protobuf/lib/libabsl_throw_delegate.a  /opt/protobuf/lib/libabsl_raw_logging_internal.a  /opt/protobuf/lib/libabsl_log_severity.a && :
```

and the link command for Debug, which fails:

```
: && /usr/bin/clang++-18 -g  CMakeFiles/messages.dir/generated/src/messages.pb.cc.o CMakeFiles/main.dir/src/main.cpp.o -o main  /opt/protobuf/lib/libprotobuf.a  /opt/protobuf/lib/libabsl_log_internal_check_op.a  /opt/protobuf/lib/libabsl_leak_check.a  /opt/protobuf/lib/libabsl_die_if_null.a  /opt/protobuf/lib/libabsl_log_internal_conditions.a  /opt/protobuf/lib/libabsl_log_internal_message.a  /opt/protobuf/lib/libabsl_log_internal_nullguard.a  /opt/protobuf/lib/libabsl_examine_stack.a  /opt/protobuf/lib/libabsl_log_internal_format.a  /opt/protobuf/lib/libabsl_log_internal_proto.a  /opt/protobuf/lib/libabsl_log_internal_log_sink_set.a  /opt/protobuf/lib/libabsl_log_sink.a  /opt/protobuf/lib/libabsl_log_entry.a  /opt/protobuf/lib/libabsl_flags_internal.a  /opt/protobuf/lib/libabsl_flags_marshalling.a  /opt/protobuf/lib/libabsl_flags_reflection.a  /opt/protobuf/lib/libabsl_flags_config.a  /opt/protobuf/lib/libabsl_flags_program_name.a  /opt/protobuf/lib/libabsl_flags_private_handle_accessor.a  /opt/protobuf/lib/libabsl_flags_commandlineflag.a  /opt/protobuf/lib/libabsl_flags_commandlineflag_internal.a  /opt/protobuf/lib/libabsl_log_initialize.a  /opt/protobuf/lib/libabsl_log_internal_globals.a  /opt/protobuf/lib/libabsl_log_globals.a  /opt/protobuf/lib/libabsl_vlog_config_internal.a  /opt/protobuf/lib/libabsl_log_internal_fnmatch.a  /opt/protobuf/lib/libabsl_raw_hash_set.a  /opt/protobuf/lib/libabsl_hash.a  /opt/protobuf/lib/libabsl_city.a  /opt/protobuf/lib/libabsl_low_level_hash.a  /opt/protobuf/lib/libabsl_hashtablez_sampler.a  /opt/protobuf/lib/libabsl_random_distributions.a  /opt/protobuf/lib/libabsl_random_seed_sequences.a  /opt/protobuf/lib/libabsl_random_internal_pool_urbg.a  /opt/protobuf/lib/libabsl_random_internal_randen.a  /opt/protobuf/lib/libabsl_random_internal_randen_hwaes.a  /opt/protobuf/lib/libabsl_random_internal_randen_hwaes_impl.a  /opt/protobuf/lib/libabsl_random_internal_randen_slow.a  /opt/protobuf/lib/libabsl_random_internal_platform.a  /opt/protobuf/lib/libabsl_random_internal_seed_material.a  /opt/protobuf/lib/libabsl_random_seed_gen_exception.a  /opt/protobuf/lib/libabsl_statusor.a  /opt/protobuf/lib/libabsl_status.a  /opt/protobuf/lib/libabsl_cord.a  /opt/protobuf/lib/libabsl_cordz_info.a  /opt/protobuf/lib/libabsl_cord_internal.a  /opt/protobuf/lib/libabsl_cordz_functions.a  /opt/protobuf/lib/libabsl_exponential_biased.a  /opt/protobuf/lib/libabsl_cordz_handle.a  /opt/protobuf/lib/libabsl_crc_cord_state.a  /opt/protobuf/lib/libabsl_crc32c.a  /opt/protobuf/lib/libabsl_crc_internal.a  /opt/protobuf/lib/libabsl_crc_cpu_detect.a  /opt/protobuf/lib/libabsl_bad_optional_access.a  /opt/protobuf/lib/libabsl_strerror.a  /opt/protobuf/lib/libabsl_str_format_internal.a  /opt/protobuf/lib/libabsl_synchronization.a  /opt/protobuf/lib/libabsl_stacktrace.a  /opt/protobuf/lib/libabsl_symbolize.a  /opt/protobuf/lib/libabsl_debugging_internal.a  /opt/protobuf/lib/libabsl_demangle_internal.a  /opt/protobuf/lib/libabsl_graphcycles_internal.a  /opt/protobuf/lib/libabsl_kernel_timeout_internal.a  /opt/protobuf/lib/libabsl_malloc_internal.a  /opt/protobuf/lib/libabsl_time.a  /opt/protobuf/lib/libabsl_civil_time.a  /opt/protobuf/lib/libabsl_time_zone.a  /opt/protobuf/lib/libabsl_bad_variant_access.a  /opt/protobuf/lib/libutf8_validity.a  /opt/protobuf/lib/libabsl_strings.a  /opt/protobuf/lib/libabsl_int128.a  /opt/protobuf/lib/libabsl_strings_internal.a  /opt/protobuf/lib/libabsl_string_view.a  /opt/protobuf/lib/libabsl_base.a  /opt/protobuf/lib/libabsl_spinlock_wait.a  -lrt  /opt/protobuf/lib/libabsl_throw_delegate.a  /opt/protobuf/lib/libabsl_raw_logging_internal.a  /opt/protobuf/lib/libabsl_log_severity.a && :
```

When I diff these, the only difference is with these flags:

```diff
< -O3
< -DNDEBUG
---
> -g
```

**Comment by e8y** [comment]
This seems to be a matter of inconsistency with compiling protobuf and abseil with the system-default compiler and linking it with the application's compiler toolchain (LLVM). If I use the same toolchain for both protobuf and the application (either the system one or the custom-installed LLVM one), this works fine.

**Comment by tyler92** [comment]
@e8y Did you find a solution?

**Comment by justusranvier** [comment]
Apparently a shared library which uses protoc generated classes internally, even if none of those symbols are visible outside the shared library, under circumstances can not be linked to a completely different application due to non-obvious toolchain incompatibilities.

I encountered this issue when I tried to make a debug build of an application compiled with `-O0 ` which links to a shared library which uses protobuf generated types internally without exporting their symbols.

Any optimization level above `-O0` worked, but a completely un-optimized build failed.
